### DISCUSSION TITLE
Dynamic memory

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2164] [Q&A] [answered]
I need to use dynamic memory in my research and I'm not sure which option to use. Thank you in advance for thinking about my questions below. I have a use case that need to remember `n` things as they occur during runtime and `n` isn't known.

1) Option 1, use `MEMB`, but I'm puzzled by advantage over a standard fixed size array. `MEMB` API require that the number of objects will be specified, so what is the difference between it and a regular array?

2) `HeapMem` is well justified in the new [docs](https://contiki-ng.readthedocs.io/en/develop/doc/programming/Memory-management.html#heap-memory-heapmem) but I didn't find a single use of it in Contiki-NG source, although I'm certain there were use cases that could benefit from it.

Is it safe to use?

If the objects I dynamically create and store are `struct` of a few types will I be able to cast the pointer?

Will I be able to use the standard C `struct a1 = a2` to copy structures of the same type?

**Comment by nvt** [comment, marked as answer]
1. MEMB keeps track of which array items are in use, and provides functionality to allocate and deallocate items.

2. HeapMem has been used in several projects using Contiki-NG, but the software is not in the main repository.

The minimum alignment can be configured through the HEAPMEM_CONF_ALIGNMENT parameter. Then the memory alignment will also depend on the particular size of the internal "chunk" structure of the module. To be most safe, you can use memcpy the structure from and to the heap rather than casting a structure pointer into the allocated memory, even though the latter should work in most cases.

There is now a test for the HeapMem module in tests/08-native-runs/12-heapmem that can be executed by running the script 12-heapmem.sh in the directory one level above.

**Reply by rahav** [reply]
Thanks for the fast response!!! Both MEMB and HeapMem sound great features to have.

**Comment by pjonsson** [comment]
It's probably not what you were looking for, but I'll mention variable length arrays for completeness of the toolbox. Example without input validation on Linux:

```C
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[]) {
  int arr[atoi(argv[1])];
  printf("array size: %zu bytes\n", sizeof(arr));
  printf("number of objects: %zu\n", sizeof(arr)/sizeof(arr[0]));
  return 0;
}
```

```
$ clang-14 -Wall test.c
$ ./a.out 3
array size: 12 bytes
number of objects: 3
```

**Reply by rahav** [reply]
Thank you Peter, this could be handy. Any advise on when to use MEMB or HeapMem?
